Improve performance of crc32_acle on 32-bit ARM

ccawley2011 opened the pull request:
This does a few things:
- `__crc32d` is called even on AArch32, where it expands to two `__crc32w` instructions.
- ~The number of intermediate variables is reduced, which makes modern GCC more likely to use `ldrd` or `ldm` instead of two `ldr` instructions.~ This wasn't as beneficial as I'd hoped.
- ~`size_t` is used for the length to avoid 64-bit comparisons and subtractions in the inner loop. This can be simplified once PR #1356 is merged.~
- The alignment checks are simplified so that most of them are skipped if the buffer is already aligned.
- A special case is added when the length is 1.


Results from `benchmark_zlib` on a Raspberry Pi 400 (outdated) - Before:
```
----------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations
----------------------------------------------------------------------------------
crc32/acle/1                                  6.22 ns         6.22 ns    112505190
crc32/acle/8                                  25.9 ns         25.9 ns     27547565
crc32/acle/64                                 37.8 ns         37.8 ns     18518587
crc32/acle/512                                 235 ns          235 ns      2977104
crc32/acle/4096                               1730 ns         1729 ns       404738
crc32/acle/32768                             13968 ns        13967 ns        50177
crc32/acle/262144                           110440 ns       110436 ns         6327
crc32/acle/2097152                          984960 ns       984810 ns          707
crc32/acle/4194304                         1966942 ns      1966880 ns          355
```

After:
```
----------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations
----------------------------------------------------------------------------------
crc32/acle/1                                  11.7 ns         11.7 ns     59957221
crc32/acle/8                                  16.7 ns         16.7 ns     41973016
crc32/acle/64                                 27.8 ns         27.8 ns     25184196
crc32/acle/512                                99.0 ns         98.9 ns      7074500
crc32/acle/4096                                597 ns          597 ns      1171412
crc32/acle/32768                              5081 ns         5080 ns       134381
crc32/acle/262144                            38645 ns        38641 ns        18111
crc32/acle/2097152                          559192 ns       559120 ns         1238
crc32/acle/4194304                         1118589 ns      1118471 ns          623
```

ccawley2011 committed:
Improve performance of crc32_acle on 32-bit ARM

mtl1979 commented:
This seems to remove early exit for short lengths and that will cause net performance loss with short lengths due to added cost of while loop.

ccawley2011 commented:
> This seems to remove early exit for short lengths and that will cause net performance loss with short lengths due to added cost of while loop.

It's still there, it just happens slightly later after the 64-bit alignment check.

mtl1979 commented:
@ccawley2011 Somewhere disappears over 5.4 ns of performance and that is way more than error margin...

nmoinvaz commented:
It might be good idea to use `--benchmark_repetitions=5` to ensure results.

ccawley2011 commented:
I've managed to improve the performance with lower sizes. The original post has been updated with the new changes.

Results from `benchmark_zlib` on a Raspberry Pi 3B - Before:

```
--------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
crc32/acle/1                    19.7 ns         19.2 ns     36446469
crc32/acle/1                    19.8 ns         19.2 ns     36446469
crc32/acle/1                    19.8 ns         19.3 ns     36446469
crc32/acle/1                    19.5 ns         19.3 ns     36446469
crc32/acle/1                    19.4 ns         19.3 ns     36446469
crc32/acle/1_mean               19.7 ns         19.3 ns            5
crc32/acle/1_median             19.7 ns         19.3 ns            5
crc32/acle/1_stddev            0.159 ns        0.067 ns            5
crc32/acle/1_cv                 0.81 %          0.35 %             5
crc32/acle/8                    41.2 ns         40.9 ns     17121534
crc32/acle/8                    40.9 ns         40.9 ns     17121534
crc32/acle/8                    42.3 ns         41.0 ns     17121534
crc32/acle/8                    42.1 ns         41.0 ns     17121534
crc32/acle/8                    42.3 ns         41.0 ns     17121534
crc32/acle/8_mean               41.8 ns         41.0 ns            5
crc32/acle/8_median             42.1 ns         41.0 ns            5
crc32/acle/8_stddev            0.656 ns        0.064 ns            5
crc32/acle/8_cv                 1.57 %          0.16 %             5
crc32/acle/64                    100 ns         99.7 ns      7037003
crc32/acle/64                    100 ns         99.7 ns      7037003
crc32/acle/64                   99.6 ns         99.3 ns      7037003
crc32/acle/64                   99.5 ns         99.3 ns      7037003
crc32/acle/64                    104 ns         99.6 ns      7037003
crc32/acle/64_mean               101 ns         99.5 ns            5
crc32/acle/64_median             100 ns         99.6 ns            5
crc32/acle/64_stddev            1.75 ns        0.198 ns            5
crc32/acle/64_cv                1.74 %          0.20 %             5
crc32/acle/512                   482 ns          474 ns      1472351
crc32/acle/512                   473 ns          473 ns      1472351
crc32/acle/512                   473 ns          473 ns      1472351
crc32/acle/512                   481 ns          475 ns      1472351
crc32/acle/512                   473 ns          473 ns      1472351
crc32/acle/512_mean              476 ns          473 ns            5
crc32/acle/512_median            473 ns          473 ns            5
crc32/acle/512_stddev           4.58 ns        0.826 ns            5
crc32/acle/512_cv               0.96 %          0.17 %             5
crc32/acle/4096                 3463 ns         3463 ns       202146
crc32/acle/4096                 3478 ns         3473 ns       202146
crc32/acle/4096                 3463 ns         3463 ns       202146
crc32/acle/4096                 3465 ns         3464 ns       202146
crc32/acle/4096                 3466 ns         3465 ns       202146
crc32/acle/4096_mean            3467 ns         3465 ns            5
crc32/acle/4096_median          3465 ns         3464 ns            5
crc32/acle/4096_stddev          6.31 ns         4.39 ns            5
crc32/acle/4096_cv              0.18 %          0.13 %             5
crc32/acle/32768               27782 ns        27782 ns        24939
crc32/acle/32768               27467 ns        27463 ns        24939
crc32/acle/32768               27497 ns        27494 ns        24939
crc32/acle/32768               27996 ns        27957 ns        24939
crc32/acle/32768               27865 ns        27863 ns        24939
crc32/acle/32768_mean          27722 ns        27712 ns            5
crc32/acle/32768_median        27782 ns        27782 ns            5
crc32/acle/32768_stddev          232 ns          222 ns            5
crc32/acle/32768_cv             0.84 %          0.80 %             5
crc32/acle/262144             221736 ns       221694 ns         3157
crc32/acle/262144             222579 ns       222302 ns         3157
crc32/acle/262144             221827 ns       221828 ns         3157
crc32/acle/262144             221807 ns       221795 ns         3157
crc32/acle/262144             222039 ns       221999 ns         3157
crc32/acle/262144_mean        221998 ns       221924 ns            5
crc32/acle/262144_median      221827 ns       221828 ns            5
crc32/acle/262144_stddev         344 ns          238 ns            5
crc32/acle/262144_cv            0.15 %          0.11 %             5
crc32/acle/2097152           1947559 ns      1944459 ns          360
crc32/acle/2097152           2030000 ns      2001681 ns          360
crc32/acle/2097152           1932052 ns      1932075 ns          360
crc32/acle/2097152           1941378 ns      1941300 ns          360
crc32/acle/2097152           1920761 ns      1920105 ns          360
crc32/acle/2097152_mean      1954350 ns      1947924 ns            5
crc32/acle/2097152_median    1941378 ns      1941300 ns            5
crc32/acle/2097152_stddev      43482 ns        31507 ns            5
crc32/acle/2097152_cv           2.22 %          1.62 %             5
crc32/acle/4194304           3862149 ns      3861498 ns          184
crc32/acle/4194304           3833238 ns      3832815 ns          184
crc32/acle/4194304           3831548 ns      3831588 ns          184
crc32/acle/4194304           3885858 ns      3884524 ns          184
crc32/acle/4194304           3815172 ns      3814995 ns          184
crc32/acle/4194304_mean      3845593 ns      3845084 ns            5
crc32/acle/4194304_median    3833238 ns      3832815 ns            5
crc32/acle/4194304_stddev      28155 ns        27672 ns            5
crc32/acle/4194304_cv           0.73 %          0.72 %             5
```

After:
```
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
crc32/acle/1                    14.2 ns         14.2 ns     49147254
crc32/acle/1                    14.3 ns         14.2 ns     49147254
crc32/acle/1                    14.3 ns         14.2 ns     49147254
crc32/acle/1                    14.3 ns         14.3 ns     49147254
crc32/acle/1                    14.2 ns         14.2 ns     49147254
crc32/acle/1_mean               14.3 ns         14.2 ns            5
crc32/acle/1_median             14.3 ns         14.2 ns            5
crc32/acle/1_stddev            0.033 ns        0.027 ns            5
crc32/acle/1_cv                 0.23 %          0.19 %             5
crc32/acle/8                    28.6 ns         28.5 ns     24593408
crc32/acle/8                    28.5 ns         28.4 ns     24593408
crc32/acle/8                    28.5 ns         28.4 ns     24593408
crc32/acle/8                    28.5 ns         28.5 ns     24593408
crc32/acle/8                    28.6 ns         28.6 ns     24593408
crc32/acle/8_mean               28.6 ns         28.5 ns            5
crc32/acle/8_median             28.5 ns         28.5 ns            5
crc32/acle/8_stddev            0.061 ns        0.066 ns            5
crc32/acle/8_cv                 0.21 %          0.23 %             5
crc32/acle/64                   54.0 ns         53.8 ns     13006305
crc32/acle/64                   55.6 ns         55.5 ns     13006305
crc32/acle/64                   54.2 ns         54.1 ns     13006305
crc32/acle/64                   55.1 ns         55.0 ns     13006305
crc32/acle/64                   55.1 ns         55.0 ns     13006305
crc32/acle/64_mean              54.8 ns         54.7 ns            5
crc32/acle/64_median            55.1 ns         55.0 ns            5
crc32/acle/64_stddev           0.679 ns        0.716 ns            5
crc32/acle/64_cv                1.24 %          1.31 %             5
crc32/acle/512                   303 ns          302 ns      2321625
crc32/acle/512                   302 ns          301 ns      2321625
crc32/acle/512                   302 ns          301 ns      2321625
crc32/acle/512                   302 ns          302 ns      2321625
crc32/acle/512                   303 ns          302 ns      2321625
crc32/acle/512_mean              302 ns          302 ns            5
crc32/acle/512_median            302 ns          302 ns            5
crc32/acle/512_stddev          0.464 ns        0.554 ns            5
crc32/acle/512_cv               0.15 %          0.18 %             5
crc32/acle/4096                 2181 ns         2175 ns       321630
crc32/acle/4096                 2178 ns         2173 ns       321630
crc32/acle/4096                 2186 ns         2180 ns       321630
crc32/acle/4096                 2180 ns         2179 ns       321630
crc32/acle/4096                 2182 ns         2176 ns       321630
crc32/acle/4096_mean            2181 ns         2177 ns            5
crc32/acle/4096_median          2181 ns         2176 ns            5
crc32/acle/4096_stddev          3.26 ns         3.07 ns            5
crc32/acle/4096_cv              0.15 %          0.14 %             5
crc32/acle/32768               17682 ns        17676 ns        40180
crc32/acle/32768               17483 ns        17455 ns        40180
crc32/acle/32768               17626 ns        17622 ns        40180
crc32/acle/32768               17709 ns        17688 ns        40180
crc32/acle/32768               17631 ns        17613 ns        40180
crc32/acle/32768_mean          17626 ns        17611 ns            5
crc32/acle/32768_median        17631 ns        17622 ns            5
crc32/acle/32768_stddev         87.3 ns         93.1 ns            5
crc32/acle/32768_cv             0.50 %          0.53 %             5
crc32/acle/262144             140286 ns       140195 ns         5015
crc32/acle/262144             140659 ns       140570 ns         5015
crc32/acle/262144             140501 ns       140429 ns         5015
crc32/acle/262144             139806 ns       139797 ns         5015
crc32/acle/262144             139550 ns       139453 ns         5015
crc32/acle/262144_mean        140161 ns       140089 ns            5
crc32/acle/262144_median      140286 ns       140195 ns            5
crc32/acle/262144_stddev         469 ns          460 ns            5
crc32/acle/262144_cv            0.33 %          0.33 %             5
crc32/acle/2097152           1280425 ns      1278922 ns          544
crc32/acle/2097152           1275545 ns      1274610 ns          544
crc32/acle/2097152           1264960 ns      1264080 ns          544
crc32/acle/2097152           1268613 ns      1267873 ns          544
crc32/acle/2097152           1272909 ns      1271378 ns          544
crc32/acle/2097152_mean      1272491 ns      1271373 ns            5
crc32/acle/2097152_median    1272909 ns      1271378 ns            5
crc32/acle/2097152_stddev       6004 ns         5764 ns            5
crc32/acle/2097152_cv           0.47 %          0.45 %             5
crc32/acle/4194304           2544871 ns      2542276 ns          272
crc32/acle/4194304           2679410 ns      2675505 ns          272
crc32/acle/4194304           2531671 ns      2528941 ns          272
crc32/acle/4194304           2514319 ns      2511261 ns          272
crc32/acle/4194304           2641504 ns      2639422 ns          272
crc32/acle/4194304_mean      2582355 ns      2579481 ns            5
crc32/acle/4194304_median    2544871 ns      2542276 ns            5
crc32/acle/4194304_stddev      73350 ns        73154 ns            5
crc32/acle/4194304_cv           2.84 %          2.84 %             5
```

mtl1979 commented:
I'm generally in favor of improving the optimized code, but in this case we need to postpone until #1356 is merged so we don't need to introduce the static function and dummy while loop or implicit truncating cast to avoid overflow warnings on systems where `uint64_t` requires two registers and we want to enable all optional warnings.

ccawley2011 commented:
> I'm generally in favor of improving the optimized code, but in this case we need to postpone until #1356 is merged so we don't need to introduce the static function and dummy while loop or implicit truncating cast to avoid overflow warnings on systems where `uint64_t` requires two registers and we want to enable all optional warnings.

That PR is merged now, so I've rebased this PR and removed the wrapper function.

nmoinvaz commented on arch/arm/crc32_acle.c:
```diff
@@ -17,34 +17,40 @@ uint32_t crc32_acle(uint32_t crc, const uint8_t *buf, size_t len) {
-        buf4 = (const uint32_t *) buf;
-    }
 
-#if defined(__aarch64__) || defined(_M_ARM64)
```
Is it okay that this #ifdef was removed?

mtl1979 replied:
The issue isn't necessary removing it, but that the code below for AArch64 also changes, which is unacceptable.

mtl1979 replied:
Some compilers might not implement the 64-bit wide intrinsics for 32-bit architectures, but that's what for we test 32-bit ARM with both Windows and Linux compilers.

I'm not sure when Microsoft adds ACLE support to the 32-bit ARM compiler, but until that happens, crc32 code is completely disabled in nmake Makefile.

```
---------- INTRIN.H
__MACHINEARM64(unsigned __int32 __crc32b(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32h(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32w(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32d(unsigned __int32, unsigned __int64))
__MACHINEARM64(unsigned __int32 __crc32cb(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32ch(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32cw(unsigned __int32, unsigned __int32))
__MACHINEARM64(unsigned __int32 __crc32cd(unsigned __int32, unsigned __int64))
__MACHINEX86_X64(unsigned int _mm_crc32_u16(unsigned int, unsigned short))
__MACHINEX86_X64(unsigned int _mm_crc32_u32(unsigned int, unsigned int))
__MACHINEX64(unsigned __int64 _mm_crc32_u64(unsigned __int64, unsigned __int64))
__MACHINEX86_X64(unsigned int _mm_crc32_u8(unsigned int, unsigned char))
```

mtl1979 replied:
Instead of checking if the architecture is 32-bit or 64-bit, we should only check that we're not using `_M_ARM` as that definitely doesn't support `__crc32d` and is likely to not support when ARMv8 is added as valid target.

mtl1979 commented on arch/arm/crc32_acle.c:
```diff
@@ -17,34 +17,40 @@ uint32_t crc32_acle(uint32_t crc, const uint8_t *buf, size_t len) {
+    } else {
+        buf8 = (const uint64_t *) buf;
+    }
 
```
If `len` is already 0 here, we should bail out early.

nmoinvaz replied:
An extra if statement will need to be added, that will be hit for all paths. Where as if `len = 0` then it only hits an extra `if` statement at the end and maybe an assignment operator. I think we should leave it as is.

Dead2 commented:
Awaiting @ccawley2011 replies to reviews.

nmoinvaz commented:
> __crc32d is called even on AArch32, where it expands to two __crc32w instructions.

I think after all the code changes we have had since this PR opened, that perhaps this is a good approach:

* Rename `neon_intrins.h` to `arm_intrins.h` and update build systems
* Add static inline `__crc32d` shim to `arm_intrins.h` that calls two `__crc32w` if AArch32 is detected. 

This will make it less if/def'y.

nmoinvaz commented:
Perhaps something like this in `arm_intrins.h`:
```c
#ifdef ARM_ACLE

#if !defined(__aarch64__) && !defined(_M_ARM64) && !defined(_M_ARM64EC)
#  define __crc32d(a, b) __crc32w(__crc32w(a, (uint32_t)(b)), (uint32_t)((b) >> 32));
#endif

#endif
```

ccawley2011 commented:
> * Rename `neon_intrins.h` to `arm_intrins.h` and update build systems

It would probably be best to have two separate headers to avoid issues with including NEON headers and inline functions in non-NEON files, which was one of the main reasons I had for splitting it out from `fallback_intrins.h` originally. PR #1538 introduces a separate header for ACLE intrinsics, so once that's merged I'll adapt the CRC32 code to match.

> * Add static inline `__crc32d` shim to `arm_intrins.h` that calls two `__crc32w` if AArch32 is detected.

I can do that, but I'm not aware of any AArch32 compilers that support `__crc32w` but not `__crc32d`.

nmoinvaz commented:
> I can do that, but I'm not aware of any AArch32 compilers that support __crc32w but not __crc32d.

If that is the case, then don't worry about it. I think we can cross that bridge when we come to it.

nmoinvaz commented:
> I can do that, but I'm not aware of any AArch32 compilers that support __crc32w but not __crc32d.

The only reason I suggest it, is that the original code does not appear to call `__crc32d` for AArch32. If it becomes overly complex to implement my suggestion, then we can leave it out and see if anybody comes across the issue on AArch32 in the future.

nmoinvaz commented:
Some CI was failing at the time, it might be good idea to rebase so we can see why. GH purged all the logs.

Dead2 commented:
@ccawley2011 Sorry, I attempted to resolve the merge conflicts, but I botched and committed it without being able to see the actual changes due to githubs horrible interface for merge conflict resolution. And of course it won't let me correct it now that there are no conflicts anymore.

Would you please do a rebase and force push?
